elements/lisk-tree/src/sparse_merkle_tree/sparse_merkle_tree.ts

```diff
@@ -35,8 +43,108 @@ export class SparseMerkleTree {
 	public get db(): Database {
 		return this._db;
 	}
+
+	public async getNode(nodeHash: Buffer): Promise<TreeNode> {
+		const data = await this._hashToValueMap.get(nodeHash);
+
+		if (!data) {
+			throw new Error(`Hash does not exist in merkle tree: ${nodeHash.toString('hex')}`);
+		}
+
+		if (isLeaf(data)) {
+			const { key, value } = getLeafData(data, this.keyLength);
+			return new Leaf(key, value);
+		}
+
+		const { leftHash, rightHash } = getBranchData(data);
+
+		return new Branch(leftHash, rightHash);
+	}
+
+	public async update(key: Buffer, value: Buffer) {
+		if (!value) {
+			throw new Error('Value cannot be empty');
+		}
+
+		if (key.length !== this.keyLength) {
+			throw new Error(`Key is not equal to defined key length of ${this.keyLength}`);
+		}
+
+		const newLeaf = new Leaf(key, value);
+		const ancestorNodes: TreeNode[] = [];
+		let bottomNode: TreeNode = new Empty();
+		let currentNode = await this.getNode(this.rootHash);
+		const binaryKey = binaryExpansion(key, this.keyLength);
+		let h = 0;
+
+		while (!isLeaf(currentNode.hash)) {
+			// Append currentNode to ancestorNodes
+			ancestorNodes.push(currentNode);
+			const d = binaryKey[h];
+			if (+d === 0) {
+				currentNode = await this.getNode(currentNode.hash);
+			} else if (+d === 1) {
+				currentNode = await this.getNode(currentNode.hash);
+			}
+			h += 1;
+			// The currentNode is an empty default node or a leaf node
+		}
+		// The currentNode is an empty node
+		if (currentNode.hash.equals(EMPTY_HASH)) {
+			bottomNode = newLeaf;
+		} else if (currentNode instanceof Leaf && key.equals(currentNode.key)) {
+			bottomNode = newLeaf;
+		} else {
+			// We need to create new branches in the tree to fulfill the
+			// Condition of one leaf per empty subtree
+			// Note: h is set to the last value from the previous loop
+			let d = binaryKey[h];
+			const currentNodeBinaryKey = binaryExpansion((currentNode as Leaf).key, this.keyLength);
+			let t = currentNodeBinaryKey[h];
+			while (d === t) {
+				// Create branch node with empty value
+				const defaultBranch = new Branch(EMPTY_HASH, EMPTY_HASH);
+				// Append defaultBranch to ancestorNodes
+				ancestorNodes.push(defaultBranch);
+				h += 1;
+				d = binaryKey[h];
+				t = currentNodeBinaryKey[h];
+			}
+			// Create last branch node, parent of node and newLeaf
+			if (+d === 0) {
+				bottomNode = new Branch(newLeaf.hash, currentNode.hash);
+			} else if (+d === 1) {
+				bottomNode = new Branch(currentNode.hash, newLeaf.hash);
+			}
+		}
+		// Finally update all branch nodes in ancestorNodes
+		// Starting from the last
+		while (h > 0) {
+			const d = +binaryKey[h];
+			const p = ancestorNodes[h - 1];
+			if (d === 0 && p instanceof Branch) {
```

Comment:
Is there a possibility that we have binary key value either `0` or `1` and node there in path is not a branch node?

Reply:
if you see ancestorNodes array we are only pushing branch (or default branch) nodes so there will not be any non branch nodes